engine-sdk/src/bn128.rs

```diff
@@ -0,0 +1,169 @@
+use aurora_engine_types::Vec;
+
+#[cfg(feature = "contract")]
+use super::exports;
+
+#[derive(Debug)]
+pub enum BnError {
+    Field(bn::FieldError),
+    Scalar(bn::FieldError),
+    G1(bn::GroupError),
+    G2(bn::GroupError),
+}
+
+impl From<bn::FieldError> for BnError {
+    fn from(err: bn::FieldError) -> Self {
+        Self::Field(err)
+    }
+}
+
+#[cfg(feature = "contract")]
+pub fn alt_bn128_g1_sum(left: [u8; 64], right: [u8; 64]) -> Result<[u8; 64], BnError> {
+    use aurora_engine_types::U256;
+    let mut bytes = Vec::with_capacity(64 * 2 + 2); // 64 bytes per G1 + 2 positive integer bytes.
+
+    bytes.push(0); // positive sign
+    bytes.extend_from_slice(&left);
+    bytes.push(0);
+    bytes.extend_from_slice(&right);
+
+    let value_ptr = bytes.as_ptr() as u64;
+    let value_len = bytes.len() as u64;
+
+    unsafe {
+        const REGISTER_ID: u64 = 1;
+        exports::alt_bn128_g1_sum(value_len, value_ptr, REGISTER_ID);
+        let mut output = [0u8; 64];
+        exports::read_register(REGISTER_ID, output.as_ptr() as u64);
+        let x = U256::from_little_endian(&output[0..32]);
+        let y = U256::from_little_endian(&output[32..64]);
+        output[0..32].copy_from_slice(&x.to_big_endian());
+        output[32..64].copy_from_slice(&y.to_big_endian());
+        Ok(output)
+    }
+}
+
+#[cfg(not(feature = "contract"))]
+pub fn alt_bn128_g1_sum(left: [u8; 64], right: [u8; 64]) -> Result<[u8; 64], BnError> {
+    let p1 = read_bn_g1(left)?;
+    let p2 = read_bn_g1(right)?;
+
+    let mut output = [0u8; 0x40];
```

Comment:
Nit: I think the inconsistent use of decimal and hex for length/indexing of byte arrays adds slight cognitive overhead in reading the code. I think it would be better to stick with one notation (probably decimal is easiest) throughout the whole module.